### PR TITLE
:bug: Delete unused ControlPlane InfrastructureMachineTemplates on rotation

### LINKED ISSUE #6375: ClusterClass: control plane template rotation leaks InfrastructureMachineTemplates
**What steps did you take and what happened:**

With ClusterClass, when a control plane InfrastructureMachineTemplate is rotated we don't delete the old template.

**What did you expect to happen:**
I expect old templates to be deleted - at least on a best effort basis.

**Anything else you would like to add:**

This can be reproduced by deploying a CAPD cluster with ClusterClass. After the initial deployment
the InfrastructureMachineTemplate of KCP must be modified to trigger rotations. This can be e.g. done 
by changing the template referenced in the ClusterClass or changing fields via patches.

**Environment:**

- Cluster-api version: latest
- Minikube/KIND version:
- Kubernetes version: (use `kubectl version`):
- OS (e.g. from `/etc/os-release`):

/kind bug
/area topology
[One or more /area label. See https://github.com/kubernetes-sigs/cluster-api/labels?q=area for the list of labels]

/cc @fabriziopandini

### PR DESCRIPTION
**Pull request by killianmuldoon**
Delete ControlPlane InfrastructureMachineTemplates on a best effort basis once they have been rotated during a reconciliation loop.

Signed-off-by: killianmuldoon <kmuldoon@vmware.com>

Fixes #6375
@chrischdi @fabriziopandini

### CODE REVIEW

File: internal/controllers/topology/cluster/reconcile_state.go [hunk 1]
```diff
@@ -230,6 +230,18 @@ func (r *Reconciler) reconcileControlPlane(ctx context.Context, s *scope.Scope)
 		return err
 	}
 
+	// If the controlPlane has infrastructureMachines and the InfrastructureMachineTemplate has changed on this reconcile
+	// delete the old template.
+	// This is a best effort deletion only and may leak templates if an error occurs during reconciliation.
+	// TODO:(killianmuldoon) Make a more holistic template deletion flow.
```
Comment (killianmuldoon) [inline internal/controllers/topology/cluster/reconcile_state.go:236]: I'll create an issue covering over the problems with the way we currently handle template rotation.

File: internal/controllers/topology/cluster/reconcile_state.go [hunk 2]
```diff
@@ -230,6 +230,18 @@ func (r *Reconciler) reconcileControlPlane(ctx context.Context, s *scope.Scope)
 		return err
 	}
 
+	// If the controlPlane has infrastructureMachines and the InfrastructureMachineTemplate has changed on this reconcile
+	// delete the old template.
+	// This is a best effort deletion only and may leak templates if an error occurs during reconciliation.
```
Comment (JoelSpeed) [inline internal/controllers/topology/cluster/reconcile_state.go:235]: Are there any reasons to keep a revision history here? I think some of the core types that have child/dependents like this keep the last few around and then GC them only once they are a certain age/are a number of revisions out of date

Comment (killianmuldoon) [reply internal/controllers/topology/cluster/reconcile_state.go:235]: I don't think so - the templates here are copied from a base type which is the source of truth for the spec of the ControlPlane machine so maybe we would want some history for the underlying object. 

This change is just a best effort deletion to stop the resource leaking though - I think we need a more holistic solution to cleaning up these resources and maybe we should think about whether a revision history is something we'd like to support.

Comment (JoelSpeed) [reply internal/controllers/topology/cluster/reconcile_state.go:235]: Seems like a reasonable argument to me, feel free to resolve this

File: internal/controllers/topology/cluster/reconcile_state_test.go [hunk 3]
```diff
@@ -734,6 +734,11 @@ func TestReconcileControlPlaneInfrastructureMachineTemplate(t *testing.T) {
 			for k, v := range tt.want.InfrastructureMachineTemplate.GetLabels() {
 				g.Expect(gotInfrastructureMachineTemplate.GetLabels()).To(HaveKeyWithValue(k, v))
 			}
+
+			// If the template was rotated during the reconcile we want to make sure the old template was deleted.
+			if tt.current.InfrastructureMachineTemplate != nil && tt.current.InfrastructureMachineTemplate.GetName() != tt.desired.InfrastructureMachineTemplate.GetName() {
+				g.Expect(r.Client.Get(ctx, client.ObjectKey{Namespace: tt.current.InfrastructureMachineTemplate.GetNamespace(), Name: tt.current.InfrastructureMachineTemplate.GetName()}, &unstructured.Unstructured{}))
```
Comment (fabriziopandini) [inline internal/controllers/topology/cluster/reconcile_state_test.go:740]: If I'm not wrong this is missing .To(something)
Also, might be to capture the error and check the error is object not found

File: internal/controllers/topology/cluster/reconcile_state.go (hunk 1 above)
Comment (sbueringer) [reply internal/controllers/topology/cluster/reconcile_state.go:236]: Maybe we shouldn't have this todo here (when we merge this PR). I think it's better to just have an issue describing the holistic problem vs having the TODO here specifically (as the problem is much more widespread then just this location)

File: internal/controllers/topology/cluster/reconcile_state.go [hunk 4]
```diff
@@ -230,6 +230,18 @@ func (r *Reconciler) reconcileControlPlane(ctx context.Context, s *scope.Scope)
...
+	// TODO:(killianmuldoon) Make a more holistic template deletion flow.
+	if s.Blueprint.HasControlPlaneInfrastructureMachine() && s.Current.ControlPlane.InfrastructureMachineTemplate != nil {
+		if s.Current.ControlPlane.InfrastructureMachineTemplate.GetName() != s.Desired.ControlPlane.InfrastructureMachineTemplate.GetName() {
+			if err := r.Client.Delete(ctx, s.Current.ControlPlane.InfrastructureMachineTemplate); err != nil {
+				return errors.Wrapf(err, "failed to delete old template %s", tlog.KObj{Obj: s.Desired.ControlPlane.Object})
```
Comment (sbueringer) [inline internal/controllers/topology/cluster/reconcile_state.go:240]:
```suggestion
				return errors.Wrapf(err, "failed to delete old template %s", tlog.KObj{Obj: s.Current.ControlPlane.InfrastructureMachineTemplate})
```
Not sure what the intention is but I think this error should contain the name of the template we couldn't delete, if we also want to have the control plane in the error message I think we should add both.

=> I think both would be ideal

**Review by sbueringer — COMMENTED** [review]
Overall looks good. If not already done, I think one manual test would be good to ensure we're not missing anything. (just create a classy cluster and rotate the inframachinetemplate of the control plane once + verify the old one is gone)

**Comment by ykakarap** [comment]
@fabriziopandini @sbueringer Addressed the PR comments.

**Comment by chrischdi** [comment]
/retest

Looks like this was a network related issue to github while fetching manifests

**Comment by sbueringer** [comment]
Thx!

/lgtm

**Comment by fabriziopandini** [comment]
/lgtm
/approve
@killianmuldoon could you kindly open an issue to track the need to find a reentrant solution for CP machines template cleanup

**Pull request merged**